freelist.go

```diff
@@ -22,6 +22,14 @@ func newFreelist() *freelist {
 	}
 }
 
+func newFreelistWithIDs(ids []pgid) *freelist {
```

Comment:
some indexing assumptions might be violated by passing in a raw pgid slice, how about:
```go
func (f *freelist) readIds(ids []pgid) {
    f.ids = ids
    ...
}
```
where `freelist.read()`'s tail calls into `readIds()`